lightning/src/routing/router.rs

```diff
@@ -876,40 +877,115 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+
+						// If the index is 0, then its the last hop in the RoutHint. The
+						// aggregate_next_hops_fee_msat for this hop was 0 but from here on out
+						// it needs to be increased beyond the value to be routed to account
+						// for the fee charged at hops and value to send from payer to payee.
+						if idx == 0 {
+							aggregate_next_hops_fee_msat = match aggregate_next_hops_fee_msat.checked_add(final_value_msat) {
+							Some(value) => value,
+							None => u64::max_value()
+							};
+						}
+					} else {
+						// If this hop was not used then there is no use checking the preceeding hops
+						// in the RouteHint. We can break by just searching for a direct channel between
+						// last checked hop and first_hop_targets
+						if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat, _)) = first_hop_targets.get(&prev_hop) {
+							add_entry!(first_hop, *our_node_id , prev_hop, dummy_directional_info, Some(outbound_capacity_msat / 1000), features, aggregate_next_hops_fee_msat, path_value_msat, aggregate_next_hops_path_htlc_minimum_msat);
+						}
+						break;
 					}
+
+					if idx == route.0.len() - 1 {
+						// The last hop in this iterator is the first hop in
+						// overall RouteHint.
+						// If this hop connects to a node with which we have a direct channel,
+						// ignore the network graph and, if the last hop was added, add our
+						// direct channel to the candidate set.
+						//
+						// Note that we *must* check if the last hop was added as `add_entry`
+						// always assumes that the third argument is a node to which we have a
+						// path.\
+						if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat, _)) = first_hop_targets.get(&hop.src_node_id) {
+							add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, Some(outbound_capacity_msat / 1000), features, aggregate_next_hops_fee_msat, path_value_msat, aggregate_next_hops_path_htlc_minimum_msat);
+						}
+					}
+
+					aggregate_next_hops_fee_msat = match aggregate_next_hops_fee_msat.checked_add(match idx {
+
+						// In the next values of the iterator, the aggregate fees already reflects
+						// the sum of value sent from payer (final_value_msat) and routing fees
+						// for the last node in the RouteHint. We need to just add the fees to
+						// route through the current node so that the preceeding node (next iteration)
+						// can use it.
+						_ => match compute_fees(aggregate_next_hops_fee_msat, hop.fees) {
+							Some(fees_msat) => fees_msat,
+							None => 0
+							}
+					}) {
+						Some(fees_msat) => fees_msat,
+						None =>  u64::max_value()
+					};
+
+					aggregate_next_hops_path_htlc_minimum_msat = match aggregate_next_hops_path_htlc_minimum_msat.checked_add(hop_htlc_minimum_msat) {
```

Comment:
The HTLC minimum is not increased at each hop, its the minimum *at* that hop. The value from previous hops needs to be increased by the feerate to meet the minimum at that hop, but you dont add the new hop's min. The way we calculate it in the `add_entry` loop is:

```
						// Note that low contribution here (limited by available_liquidity_msat)
						// might violate htlc_minimum_msat on the hops which are next along the
						// payment path (upstream to the payee). To avoid that, we recompute path
						// path fees knowing the final path contribution after constructing it.
						let path_htlc_minimum_msat = match compute_fees($next_hops_path_htlc_minimum_msat, $directional_info.fees)
								.map(|fee_msat| fee_msat.checked_add($next_hops_path_htlc_minimum_msat)) {
							Some(Some(value_msat)) => cmp::max(value_msat, $directional_info.htlc_minimum_msat),
							_ => u64::max_value()
						};
```